### PR TITLE
✨  Fileformat v2

### PR DESCRIPTION
**Pull request by pajowu**

### CODE REVIEW

File: app/src/util/index.ts [hunk 1]
```diff
@@ -28,3 +35,7 @@ export function switchExtension(pathName: string, extension: string): string {
   }
   return path.join(dirname, basename);
 }
+
+export function isRunningInTest(): boolean {
+  return process.env.JEST_WORKER_ID !== undefined;
```
Comment (anuejn) [inline app/src/util/index.ts:40]: uiuiui

File: app/src/util/memoize.spec.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  state.cursor.playerTime = 1;
+  expect(counter.current).toBe(0);
+  memoizedTime(state);
+  expect(counter.current).toBe(1);
+  state = produce(state, (state) => {
+    state.cursor.current = 'user';
+  });
+  memoizedTime(state);
+  expect(counter.current).toBe(1);
+});
```
Comment (anuejn) [inline app/src/util/memoize.spec.ts:92]: maybe we should add a testcase that breaks with upstream but works with our vendored fork of proxy-memoize

File: app/src/pages/Editor/Document.tsx [hunk 3]
```diff
@@ -161,8 +142,11 @@ export function Document(): JSX.Element {
       e.preventDefault();
     }
   };
-  function getSpeakerColor(speaker: string) {
-    const color_idx = speakerIndices[speaker] % Object.keys(theme.colors.speakers).length;
+  function getSpeakerColor(speaker: string | null) {
+    // TODO: Handle speaker == null everywhere properly
```
[the remark this answers is not in the document]
Comment (pajowu) [reply app/src/pages/Editor/Document.tsx:146]: in another PR

File: app/src/util/index.ts (hunk 1 above)
Comment (anuejn) [reply app/src/util/index.ts:40]: TODO: fix jest warnings (maybe sometime else)

File: app/src/state/editor/selection.ts [hunk 4]
```diff
@@ -10,134 +20,180 @@ export const setSelection = createActionWithReducer<EditorState, Selection | nul
...
-  const getItemLeft = (time: number) =>
-    DocumentGenerator.fromParagraphs(state.document.content).getItemsAtTime(time)[0];
-  if (!selectionInfo || !state.selection) {
-    const item = getItemLeft(state.currentTimePlayer);
-    assertSome(item);
-    state.selection = {
-      range: { start: item.absoluteStart, length: item.length },
-      startItem: item,
-    };
+function changeSelectionLeft(state: EditorState) {
```
[the remark this answers is not in the document]
Comment (pajowu) [reply app/src/state/editor/selection.ts:23]: changeSelectionHeadLeft

File: app/src/pages/Editor/Document.tsx [hunk 5]
```diff
@@ -105,32 +93,21 @@ export function Document(): JSX.Element {
...
-      const range = {
-        start: startItem.absoluteStart,
-        length:
-          endItem.absoluteStart +
-          (selectionRange.endOffset == 0 ? 0 : endItem.length) -
-          startItem.absoluteStart,
+      const range: Selection = {
+        startIndex: startItem.absoluteIndex,
+        length: endItem.absoluteIndex - startItem.absoluteIndex + 1,
+        // TODO: calculate this from selection anchor & focus node
```
[the remark this answers is not in the document]
Comment (anuejn) [reply app/src/pages/Editor/Document.tsx:115]: Is this really fixed?

[the remark this answers is not in the document]
Comment (pajowu) [reply app/src/pages/Editor/Document.tsx:115]: I opened an issue #265 for this

**Review by anuejn — APPROVED** [review]

**Comment by anuejn** [comment]
party parrot

**Comment by anuejn** [comment]
thanks for the really good and hard work :)

**Pull request merged**